contributors/design-proposals/network/service-topology-key.md

```diff
@@ -0,0 +1,103 @@
+
+## Non-goal
+
+## Use cases:
+
+* Logging agents such as fluentd. Deploy fluentd as DaemonSet and applications only need to communicate with the fluentd in the same node.
+* For a sharded service that keeps per-node local information in each shard.
+* Authenticating proxies such as [aws-es-proxy](https://github.com/kopeio/aws-es-proxy).
+* In container identity wg, being able to give daemonset pods a unique identity per host is on the 2018 plan, and ensuring local pods can communicate to local node services securely is a key goal there. -- from @smarterclayton
+* Regional data costs in multi-AZ setup - for instance, in AWS, with a multi-AZ setup, half of the traffic will switch AZ, incurring regional data Transfer costs, whereas if something was local, it wouldn't hit the network.
+* Users want a priority-based topology aware routing  service. Say, try to visit a "host-local" service backend firstly. If no available backend, then try a "zone-local" service backend...
+
+## Background
+
+It has been a very often requested feature because people need to create a service which connects to local backend Pods, the so called "only-local" service. In this way, they can reduce network latency, improve security, save money and so on. However, because topology is arbitrary, zone, region, rack, generator, whatever, who knows? We should allow arbitrary locality.
+
+`ExternalTrafficPolicy` was added in v1.4, but only for NodePort and external LB traffic. NodeName was added to `EndpointAddress` to allow kube-proxy to filter local endpoints for various future purposes.
+
+Based on our experience of advanced routing setup and recent demo of enabling this in Kubernetes service, this document would like to introduce a more generic way to enable "local-only service".
+
+## Proposal
+
+This proposal builds off of earlier requests to [use local pods only for kube-proxy loadbalancing](https://github.com/kubernetes/kubernetes/issues/7433) and [node-local service proposal](https://github.com/kubernetes/kubernetes/pull/28637). But, this document proposes that not only the particular "node-local" user case should be taken care, but also a more generic way should be figured out.
+
+## Implementation details
+
+### API changes
+
+```go
+type Service struct {
+  spec:
+    // topologyKey is the key for the node label that the system uses to denote such a topology domain. There are some built-in topology keys, e.g. kubernetes.io/hostname,
+    // failure-domain.beta.kubernetes.io/zone,
+    // failure-domain.beta.kubernetes.io/region,
+    // TopologyKeyList is a topologyKey list, which is used to guarantee ability. For example,
+    // `TopologyKeyList: [ "kubernetes.io/hostname", "failure-domain.beta.kubernetes.io/zone" ]`
+    // If failed to match hostname, then it will try to find the endpoints in the same zone
+    // The priority of key in TopologyKeyList is sorted by index and selecting process will stop once a key matches.
+  	TopologyKeyList []string
```

Comment:
As per email, we should probably be consistent with pod affinity, and make this a singular entry.  Maybe:

```
TopologyKey string
TopologyMode string  // "Ignored" (default), "Preferred", "Required"
```

Then we can see if there is really a need to make it a list or be more complicated.

Reply:
What happens if Mode is Required and no endpoints satisfy that, how do we surface this error ?

Reply:
I followed the "Required/Preferred" pattern in podAffinity, PTAL.

Reply:
> What happens if Mode is Required and no endpoints satisfy that, how do we surface this error ?

If "hard" requirement is not met, kube-proxy will not pick any endpoints for service. Users visit service that without real backends will get a "connection refused" error.